src/app/clusters/service-area-server/service-area-delegate.h

```diff
@@ -0,0 +1,388 @@
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "service-area-cluster-objects.h"
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ServiceArea {
+
+class Instance;
+
+// *****************************************************************************
+// cluster constraints
+
+constexpr size_t kMaxNumSupportedLocations = 255;
+constexpr size_t kMaxNumSupportedMaps      = 255;
+constexpr size_t kMaxNumSelectedLocations  = 255;
+constexpr size_t kMaxNumProgressElements   = 255;
+
+constexpr size_t kMaxSizeStatusText = 256;
+
+/**
+ * ServiceArea::Delegate Defines methods for implementing application-specific
+ * logic for the Service Area Cluster.
+ */
+class Delegate
+{
+public:
+    Delegate()          = default;
+    virtual ~Delegate() = default;
+
+    /**
+     * Stop this class objects from being copied.
+     */
+    Delegate(const Delegate &)             = delete;
+    Delegate & operator=(const Delegate &) = delete;
+
+    friend class Instance;
+
+protected:
+    /**
+     * @brief This method will be called during the ServiceArea server initialization after the Instance information has been
+     * validated and the Instance has been registered. This can be used to initialise app logic.
+     */
+    virtual CHIP_ERROR Init() { return CHIP_NO_ERROR; };
+
+    //*************************************************************************
+    // Command handling support
+
+    /**
+     * @brief Can the selected locations be set by the client in the current operating mode?
+     * @param[out] statusText text describing why selected locations cannot be set (if return is false).
+     * @return true if the current device state allows selected locations to be set by user.
+     *
+     * @note The statusText field SHOULD indicate why the request is not allowed, given the current mode
+     *       of the device, which may involve other clusters.
+     */
+    virtual bool IsSetSelectedLocationsAllowed(MutableCharSpan statusText) = 0;
+
+    /**
+     * Given a set of locations to be set to the SelectedLocations attribute, this method should check that:
+     *  - There are no duplicates in the list. If there are duplicates, the locationStatus SHALL be set to
+     *     DuplicatedLocations and the statusText SHALL be an empty string.
+     *  - The set of locations as a whole is valid and reachable by the device. If the set of locations is invalid,
+     *     the locationStatus should be set to InvalidSet and the statusText SHALL include a vendor-defined error description.
+     *
+     * The caller of this method will ensure that all the locations in the set are valid supported locations.
+     *
+     * @param[in] req List of new selected locations.
+     * @param[out] locationStatus Success if all checks pass, error code if failure.
+     * @param[out] statusText text describing failure (see description above), size kMaxSizeStatusText + 1 byte for terminating
```

Comment:
This documentation does not match the code, in a way that will lead to writes out of bounds if someone listens to the docs.

Plus, why do we need a "terminating character"?  The whole point of using spans, which have a length, is to not need magic in-band string end markers.

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/e17c3c10085eb3ce6a9c6279512e5c2455c34a27).